src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+    /// the same general "shape", say:
+    ///
+    /// ```ascii-art
+    /// W->X->Y->Z
+    ///    |
+    ///    +->D
+    /// ```
+    ///
+    /// So an adversary is easily able to forge any such chain.
+    ///
+    /// When the trusted key is on the main branch, on the other hand:
+    ///
+    /// ```ascii-art
+    /// D->E->F
+    ///    |
+    ///    +->G
+    /// ```
+    ///
+    /// Then such chain is impossible to forge because the adversary would have to have access to
+    /// the secret key corresponding to `D` in order to validly sign `E`. Thus such chain can be
+    /// safely considered trusted.
+    pub fn check_trust<'a, I>(&self, trusted_keys: I) -> bool
+    where
+        I: IntoIterator<Item = &'a bls::PublicKey>,
+    {
+        let trusted_keys: HashSet<_> = trusted_keys.into_iter().collect();
+        let mut index = self.tree.len();
+
+        loop {
+            let (key, parent_index) = if index > 0 {
+                let block = &self.tree[index - 1];
+                (&block.key, Some(block.parent_index))
+            } else {
+                (&self.root, None)
+            };
+
+            if trusted_keys.contains(key) {
+                return true;
+            }
+
+            if let Some(next_index) = parent_index {
+                index = next_index;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /// Compare the two keys by their position in the chain. The key that is higher (closer to the
+    /// last key) is considered `Greater`. If exactly one of the keys is not in the chain, the other
+    /// one is implicitly considered `Greater`. If none are in the chain, they are considered
+    /// `Equal`.
+    pub fn cmp_by_position(&self, lhs: &bls::PublicKey, rhs: &bls::PublicKey) -> Ordering {
+        match (self.index_of(lhs), self.index_of(rhs)) {
+            (Some(lhs), Some(rhs)) => lhs.cmp(&rhs),
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (None, None) => Ordering::Equal,
+        }
+    }
+
+    /// Returns the number of blocks in the chain. This is always >= 1.
+    pub fn len(&self) -> usize {
+        1 + self.tree.len()
+    }
+
+    fn insert_block(&mut self, new_block: Block) -> usize {
+        // Find the index into `self.tree` to insert the new key at. All the keys above will be
+        // pushed upwards.
+        let insert_at = self
+            .tree
+            .iter()
+            .enumerate()
+            .skip(new_block.parent_index)
+            .find(|(_, block)| {
+                block.parent_index != new_block.parent_index || block.key >= new_block.key
+            })
+            .map(|(index, _)| index)
+            .unwrap_or(self.tree.len());
+
+        // If the key already exists in the chain, do nothing but still return success to make the
+        // `insert` operation idempotent.
```

Comment:
but the returned value will be `insert_at + 1`, which is not the position of the existing one ?
though curretnly the returned value is ignored by the caller, does it need to be corrected?

Reply:
No, it actually is used in `merge` so this can be an issue. Good catch!

Reply:
Sorry, I was wrong. It's correct the way it is. It's because `index_at` is an index into `self.tree` but the function returns the "logical" index which is the index when we consider the `root` block as well.